src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -543,20 +556,27 @@ private async Task RefreshKeyValueCollections(CancellationToken cancellationToke
                     });
                 }
 
-                IEnumerable<KeyValueChange> keyValueChanges = await _client.GetKeyValueChangeCollection(
-                    currentKeyValues, 
-                    new GetKeyValueChangeCollectionOptions
-                    {
-                        KeyFilter = changeWatcher.Key,
-                        Label = changeWatcher.Label.NormalizeNull(),
-                        RequestTracingEnabled = _requestTracingEnabled,
-                        RequestTracingOptions = _requestTracingOptions
-                    }, 
-                    cancellationToken).ConfigureAwait(false);
+                try
+                {
+                    keyValueChanges = await _client.GetKeyValueChangeCollection(
+                        currentKeyValues,
+                        new GetKeyValueChangeCollectionOptions
+                        {
+                            KeyFilter = changeWatcher.Key,
+                            Label = changeWatcher.Label.NormalizeNull(),
+                            RequestTracingEnabled = _requestTracingEnabled,
+                            RequestTracingOptions = _requestTracingOptions
+                        },
+                        cancellationToken).ConfigureAwait(false);
 
-                changeWatcher.CacheExpires = DateTimeOffset.UtcNow.Add(changeWatcher.CacheExpirationInterval);
+                    success = true;
+                }
+                finally
+                {
+                    UpdateCacheExpirationTime(changeWatcher, success);
+                }
 
-                if (keyValueChanges?.Any() == true)
+                if (keyValueChanges.Any())

```

Comment:
> > > ```
> > >             }
> > > ```
> > 
> > 
> > If anything throws in this `if` block, do we still consider the refresh a success? In other words, should we retry upon failures?
> > Refers to: src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs:584 in [e113102](https://github.com/Azure/AppConfiguration-DotnetProvider/commit/e11310251b5a7be45c46fa077ec98dc4b7477867). [](commit_id = [e113102](https://github.com/Azure/AppConfiguration-DotnetProvider/commit/e11310251b5a7be45c46fa077ec98dc4b7477867), deletion_comment = False)
> 
> We have already updated the cache expiration time before `SetData` is called. The current refresh operation will fail if `SetData` throws, but the next refresh operation will be triggered after the cache expires.

(Codeflow messes up the github comments, so I restart it here)

Yes, we updated the cache expiration time as a successful case, so no quick retry. If an exception thrown from here will cause the app to crash, I guess it's fine. Otherwise, there will be a discrepancy between what's loaded from appconfig and what the app sees. Anyway, it's not a new problem.